plugin/keyring_vault/vault_curl.cc

```diff
@@ -0,0 +1,177 @@
+}
+
+my_bool Vault_curl::init(Vault_credentials *vault_credentials)
+{
+  curl = curl_easy_init();
+  if (curl == NULL)
+  {
+    logger->log(MY_ERROR_LEVEL, "Could not create CURL session");
+    return TRUE;
+  }
+  this->token_header = "X-Vault-Token:" + (*vault_credentials)["token"];
+  this->vault_url = (*vault_credentials)["vault_url"] + "/v1/" + (*vault_credentials)["secret_mount_point"];
+  this->vault_ca = (*vault_credentials)["vault_ca"];
+  if(this->vault_ca.empty())
+  {
+    logger->log(MY_WARNING_LEVEL, "There is no vault_ca specified in keyring_vault's configuration file. "
+                                  "Please make sure that Vault's CA certificate is trusted by the machine from "
+                                  "which you intend to connect to Vault."); 
+  }
+  return FALSE;
+}
+
+my_bool Vault_curl::reset_curl_session()
+{
+  CURLcode curl_res = CURLE_OK;
+  curl_easy_reset(curl);
+  read_data_ss.str("");
+  read_data_ss.clear();
+  curl_errbuf[0] = '\0';
+  if (list != NULL)
+  {
+    curl_slist_free_all(list);
+    list = NULL;
+  }
+
+  if ((list = curl_slist_append(list, token_header.c_str())) == NULL ||
+      (list = curl_slist_append(list, "Content-Type: application/json")) == NULL ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errbuf)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_response_memory)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEDATA, reinterpret_cast<void*>(&read_data_ss))) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L)) != CURLE_OK ||
+      (vault_ca.empty() == false &&
+       (curl_res = curl_easy_setopt(curl, CURLOPT_CAINFO, vault_ca.c_str())) != CURLE_OK
+      ) ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_USE_SSL, CURLUSESSL_ALL)) != CURLE_OK)
+  {
+    logger->log(MY_ERROR_LEVEL, get_error_from_curl(curl_res).c_str());
+    return TRUE;
+  }
+  return FALSE;
+}
+
+my_bool Vault_curl::list_keys(std::string *response)
+{
+  CURLcode curl_res = CURLE_OK;
+  curl_easy_reset(curl);
+  long http_code = 0;
+
+  if (reset_curl_session() ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_URL, (vault_url + "?list=true").c_str())) != CURLE_OK ||
+      (curl_res = curl_easy_perform(curl)) != CURLE_OK ||
+      (curl_res = curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code)) != CURLE_OK)
+  {
+    logger->log(MY_ERROR_LEVEL,
+                get_error_from_curl(curl_res).c_str());
+    return TRUE;
+  }
+  if (http_code == 404)
+  {
+    *response=""; //no keys found
+    return FALSE; 
+  }
+  *response = read_data_ss.str();
+  return http_code != 200 && http_code != 202; //200 and 202 are success return code
+}
+
+my_bool Vault_curl::write_key(IKey *key, std::string *response)
```

Comment:
unfortunately interface methods coming from framework - like get_key_data_size, which I use in write_key, are not marked as const - thus I cannot mark key as const ...